File: ros2-basic/topic_simple/src/node_helloworld_py/node_helloworld_py/node_subscriber_by_helloworld_interface.py
import rclpy
from rclpy.node import Node
from node_helloworld_interface.msg import Message


class NodeHelloworld(Node):
    def __init__(self, name: str):
        super().__init__(name)
        self.subscription_ = self.create_subscription(
            Message,
            "helloworld_py_by_helloworld_interface",
            self.subscriber_callback,
            10,
        )

    def subscriber_callback(self, msg):
        self.get_logger().info(f"[{msg.name}] I heard: '{msg.message}'")


def main(args=None):
    rclpy.init(args=args)
    node = NodeHelloworld("node_topic_subscriber_helloworld_py_by_helloworld_interface")
    rclpy.spin(node)
    rclpy.shutdown()

File: ros2-basic/topic_simple/src/node_helloworld_cpp/src/node_subscriber_by_helloworld_interface.cpp
#include "rclcpp/rclcpp.hpp"
#include "node_helloworld_interface/msg/message.hpp"

class NodeHelloworld : public rclcpp::Node
{
public:
    NodeHelloworld(std::string name) : Node(name)
    {
        subscription_ = this->create_subscription<node_helloworld_interface::msg::Message>(
            "helloworld_cpp_by_helloworld_interface",
            rclcpp::QoS(rclcpp::KeepLast(10)),
            std::bind(&NodeHelloworld::subscriber_callback, this, std::placeholders::_1));
    }

private:
    void subscriber_callback(const node_helloworld_interface::msg::Message::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "[%s] I heard: '%s'", msg->name.c_str(), msg->message.c_str());
    }

private:
    rclcpp::Subscription<node_helloworld_interface::msg::Message>::SharedPtr subscription_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NodeHelloworld>("node_topic_subscriber_helloworld_cpp_by_helloworld_interface");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}